### PR TITLE
Rework DHT

### PR DESCRIPTION
**Pull request by jobevers**
Profiling suggested that calculating distance on the dht was expensive.  Investigating, I noticed that its being used as the sort key for contact lists.  Unfortunately, python sort doesn't cache the results of the `key()` output so we could potentially be making lots of distance calculations unnecessarily.

I added a new class that does cache the output and also cache's the long value of one of the input strings.

I also refactored a function in the DHT node class because holy hell I hate long functions and nested functions and have a particular loathing for long, nested functions that reference variables from an outer scope.

### CODE REVIEW

**Comment by jobevers** [comment]
The main commit in here is https://github.com/lbryio/lbry/pull/327/commits/6bde2c4d672d37d69d426f563127243f22dce46f

**Comment by jobevers** [comment]
What?  How is this build passing?

**Comment by jobevers** [comment]
Re-ran the profiler and now the top cumulative hits are inside twisted:

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.007    0.007 6169.102 6169.102 /Users/jobevers/.virtualenvs/lbry/bin/lbrynet-daemon:3(<module>)
        1    0.000    0.000 6166.518 6166.518 lbrynet/lbrynet_daemon/DaemonControl.py:38(start)
        1    0.000    0.000 6162.416 6162.416 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/internet/base.py:1197(run)
        1   11.250   11.250 6162.412 6162.412 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/internet/base.py:1202(mainLoop)
  1110341   17.963    0.000 4830.817    0.004 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/internet/selectreactor.py:93(doSelect)
  1110342 3284.784    0.003 3299.245    0.003 {select.select}
   433515    4.477    0.000 1513.610    0.003 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/python/log.py:90(callWithLogger)
   433515    5.450    0.000 1508.541    0.003 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/python/log.py:83(callWithContext)
   433515    2.083    0.000 1494.845    0.003 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/python/context.py:117(callWithContext)
   433515    2.959    0.000 1492.411    0.003 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/python/context.py:61(callWithContext)
   433514    3.023    0.000 1488.856    0.003 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/internet/selectreactor.py:147(_doReadOrWrite)
   425259   27.067    0.000 1478.456    0.003 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/internet/udp.py:221(doRead)
  1848979   31.100    0.000 1388.717    0.001 lbrynet/dht/protocol.py:99(datagramReceived)
1110341/1110340   45.089    0.000 1289.346    0.001 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/internet/base.py:790(runUntilCurrent)
3258870/3082821   70.619    0.000  817.763    0.000 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/internet/defer.py:576(_runCallbacks)
3013560/2890791   14.067    0.000  810.626    0.000 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/internet/defer.py:548(_startRunCallbacks)
  1100914   17.369    0.000  690.492    0.001 lbrynet/dht/protocol.py:254(_handleRPC)
2239573/2120840    5.565    0.000  646.822    0.000 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/internet/defer.py:435(callback)
   583453   26.353    0.000  606.800    0.001 lbrynet/dht/node.py:765(searchIteration)
  1081373   12.748    0.000  422.528    0.000 lbrynet/dht/node.py:518(findValue)
  1081373  288.095    0.000  297.344    0.000 lbrynet/dht/hashwatcher.py:24(add_requested_hash)
  1108043    4.460    0.000  290.866    0.000 lbrynet/dht/node.py:760(sortByDistance)
  1108043    3.182    0.000  285.084    0.000 lbrynet/dht/node.py:882(sort)
   773958   12.235    0.000  270.046    0.000 lbrynet/dht/protocol.py:285(_msgTimeout)
  1481024   34.705    0.000  259.461    0.000 lbrynet/dht/node.py:812(_probeContact)
195782145  184.070    0.000  249.933    0.000 lbrynet/dht/contact.py:24(__eq__)
  1108043   31.337    0.000  238.678    0.000 lbrynet/dht/node.py:887(_cacheValues)
   707233   11.237    0.000  223.216    0.000 lbrynet/dht/node.py:672(extendShortlist)
64893273/2610815  191.193    0.000  200.890    0.000 lbrynet/dht/encoding.py:52(encode)
  1848979    4.314    0.000  199.907    0.000 lbrynet/dht/encoding.py:89(decode)
  1100914    2.701    0.000  196.865    0.000 lbrynet/dht/protocol.py:260(handleResult)
 29883190   35.932    0.000  195.660    0.000 lbrynet/dht/node.py:860(to_contact)
47074608/1848979  171.801    0.000  195.296    0.000 lbrynet/dht/encoding.py:108(_decodeRecursive)
  1100914    9.966    0.000  194.164    0.000 lbrynet/dht/protocol.py:238(_sendResponse)
  1509901    4.523    0.000  179.699    0.000 lbrynet/dht/contact.py:63(_sendRPC)
  1509901   23.342    0.000  175.176    0.000 lbrynet/dht/protocol.py:56(sendRPC)
   773987    1.972    0.000  174.983    0.000 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/internet/defer.py:460(errback)
  2610443   20.849    0.000  168.840    0.000 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/internet/udp.py:254(write)
 30793970   55.809    0.000  167.624    0.000 lbrynet/dht/node.py:852(__call__)
 35273060   41.837    0.000  148.085    0.000 {method 'encode' of 'str' objects}
```

and tottime:

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
  1110342 3284.784    0.003 3299.245    0.003 {select.select}
  1081373  288.095    0.000  297.344    0.000 lbrynet/dht/hashwatcher.py:24(add_requested_hash)
64893273/2610815  191.193    0.000  200.890    0.000 lbrynet/dht/encoding.py:52(encode)
195782145  184.070    0.000  249.933    0.000 lbrynet/dht/contact.py:24(__eq__)
47074608/1848979  171.801    0.000  195.296    0.000 lbrynet/dht/encoding.py:108(_decodeRecursive)
332861119/332861118   94.128    0.000   94.944    0.000 {isinstance}
  2610443   87.332    0.000   87.332    0.000 {method 'sendto' of '_socket.socket' objects}
3258870/3082821   70.619    0.000  817.763    0.000 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/internet/defer.py:576(_runCallbacks)
  2274024   61.984    0.000   61.984    0.000 {method 'recvfrom' of '_socket.socket' objects}
 35275912   58.449    0.000   58.449    0.000 {binascii.b2a_hex}
  4442437   57.094    0.000   87.531    0.000 {_heapq.heappop}
 30793970   55.809    0.000  167.624    0.000 lbrynet/dht/node.py:852(__call__)
  4559634   53.223    0.000  133.158    0.000 lbrynet/dht/node.py:737(_addIfValid)
1110341/1110340   45.089    0.000 1289.346    0.001 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/internet/base.py:790(runUntilCurrent)
  4758487   43.005    0.000   59.588    0.000 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/internet/base.py:706(callLater)
 35273060   41.837    0.000  148.085    0.000 {method 'encode' of 'str' objects}
 35272872   41.486    0.000  106.248    0.000 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/encodings/hex_codec.py:13(hex_encode)
 67606555   40.127    0.000   40.127    0.000 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/internet/base.py:166(__lt__)
 29883190   35.932    0.000  195.660    0.000 lbrynet/dht/node.py:860(to_contact)
  5220915   35.441    0.000   42.721    0.000 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/internet/abstract.py:484(isIPAddress)
    23782   34.817    0.001   38.306    0.002 lbrynet/core/client/StreamProgressManager.py:111(needed_blobs)
  1481024   34.705    0.000  259.461    0.000 lbrynet/dht/node.py:812(_probeContact)
  1108043   31.337    0.000  238.678    0.000 lbrynet/dht/node.py:887(_cacheValues)
  1848979   31.100    0.000 1388.717    0.001 lbrynet/dht/protocol.py:99(datagramReceived)
   773227   28.065    0.000  114.020    0.000 lbrynet/dht/node.py:744(removeFromShortlist)
   425259   27.067    0.000 1478.456    0.003 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/internet/udp.py:221(doRead)
   583453   26.353    0.000  606.800    0.001 lbrynet/dht/node.py:765(searchIteration)
  2610815   25.911    0.000   61.439    0.000 lbrynet/dht/protocol.py:221(_sendNext)
  3693169   23.536    0.000   75.021    0.000 lbrynet/dht/routingtable.py:292(_kbucketIndex)
  1509901   23.342    0.000  175.176    0.000 lbrynet/dht/protocol.py:56(sendRPC)
9058794/8819143   22.829    0.000   47.980    0.000 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/internet/defer.py:290(addCallbacks)
  4814606   21.554    0.000   21.554    0.000 {method 'sort' of 'list' objects}
  2610443   20.849    0.000  168.840    0.000 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/internet/udp.py:254(write)
  4371180   19.232    0.000   19.232    0.000 lbrynet/dht/contact.py:42(<lambda>)
  1373661   18.781    0.000   44.989    0.000 {method 'index' of 'list' objects}
  1110341   17.963    0.000 4830.817    0.004 /Users/jobevers/.virtualenvs/lbry/lib/python2.7/site-packages/twisted/internet/selectreactor.py:93(doSelect)
  2610815   17.603    0.000   33.686    0.000 lbrynet/dht/msgtypes.py:22(__init__)
  1100914   17.369    0.000  690.492    0.001 lbrynet/dht/protocol.py:254(_handleRPC)
1848981/1848979   17.329    0.000  118.113    0.000 lbrynet/dht/routingtable.py:362(addContact)
 37251866   17.271    0.000   17.271    0.000 {method 'find' of 'str' objects}
```

**Comment by jobevers** [comment]
https://github.com/lbryio/lbry/pull/326 and https://github.com/lbryio/lbry/pull/325 should be merged in first.  That will make the diff on the PR much smaller as the changes in branches are also part of this PR.

File: tests/dht/testNode.py [hunk 1]
```diff
@@ -55,9 +55,6 @@ def setUp(self):
             self.cases.append((h.digest(), 5000+2*i))
             self.cases.append((h.digest(), 5001+2*i))
             #(('a', 'hello there\nthis is a test'),
-            #         ('b', unicode('jasdklfjklsdj;f2352352ljklzsdlkjkasf\ndsjklafsd')),
-            #         ('e', 123),
-            #         ('f', [('this', 'is', 1), {'complex': 'data entry'}]),
             #         ('aMuchLongerKeyThanAnyOfThePreviousOnes', 'some data'))
```
Comment (jackrobison) [inline tests/dht/testNode.py]: nit: undeleted stuff

File: tests/dht/testNode.py [hunk 2]
```diff
@@ -105,28 +102,10 @@ def testAddSelfAsContact(self):
...
-#        self.remoteNodes = []
-#        for i in range(10):
-#            remoteNode = entangled.kademlia.node.Node()
-#           remoteContact = entangled.kademlia.contact.Contact(remoteNode.id, '127.0.0.1', 91827+i, self.node._protocol)
-#           self.remoteNodes.append(remoteNode)
-#            self.node.addContact(remoteContact)
             
             
-#    def testIterativeFindNode(self):
 #        """ Ugly brute-force test to see if the iterative node lookup algorithm runs without failing """
```
Comment (jackrobison) [inline tests/dht/testNode.py]: a couple more straggler comments

File: tests/dht/testNode.py [hunk 3]
```diff
@@ -299,145 +270,71 @@ def testNodeBootStrap(self):
         self.failUnlessEqual(activeContacts, expectedResult, \
                                  "Active should only contain the closest possible contacts which were used as input for the boostrap")
     
-#    def testFindingCloserNodes(self):
 #        """ Test discovery of closer contacts""" 
 #
```
Comment (jackrobison) [inline tests/dht/testNode.py]: straggler comments

File: tests/dht/testProtocol.py [hunk 4]
```diff
@@ -193,43 +189,11 @@ def handleResult(result):
         # The list of sent RPC messages should be empty at this stage
         self.failUnlessEqual(len(self.protocol._sentMessages), 0, 'The protocol is still waiting for a RPC result, but the transaction is already done!')
 
-#    def testDatagramLargeMessageReconstruction(self):
 #        """ Tests if a large amount of data can be successfully re-constructed from multiple UDP datagrams """
```
Comment (jackrobison) [inline tests/dht/testProtocol.py]: straggler doc string and comments

**Comment by jobevers** [comment]
merged as part of https://github.com/lbryio/lbry/pull/339

**Pull request closed without merging**